### PR TITLE
Make exporter a library.

### PR DESCRIPTION
**Pull request by jeschkies**
Summary:
- Enables users to import the exporter as a library.
- Most methods become stateless.

We would like to use YACE as a library but could not import a `main` package. Thus we propose to move al required code into `pkg`. There is some room for improvement. E.g. operational configuration such as `cloudwatchConcurrency` could be consolidate in one struct.

Depends on #315

### CODE REVIEW

File: cmd/yace/main.go [hunk 1]
```diff
@@ -191,7 +133,7 @@ func main() {
 	http.HandleFunc("/metrics", func(w http.ResponseWriter, r *http.Request) {
 		if !(*decoupledScraping) {
 			newRegistry := prometheus.NewRegistry()
-			updateMetrics(newRegistry, now)
+			exporter.UpdateMetrics(config, newRegistry, now, *metricsPerQuery, *fips, *debug, *floatingTimeWindow, *labelsSnakeCase)
```
Comment (jeschkies) [inline cmd/yace/main.go:136]: A lot of parameters are passed that could be part of an operational configuration struct.

File: pkg/abstract.go [hunk 2]
```diff
@@ -13,11 +13,11 @@ import (
 )
 
 var (
-	cloudwatchSemaphore chan struct{}
-	tagSemaphore        chan struct{}
+	cloudwatchSemaphore chan struct{} = make(chan struct{}, 5)
+	tagSemaphore        chan struct{} = make(chan struct{}, 5)
```
Comment (jeschkies) [inline pkg/abstract.go:17]: I'd like to get rid off these global variables.

**Comment by thomaspeitz** [comment]
Love this pr and would like to support it. We will first merge - https://github.com/ivx/yet-another-cloudwatch-exporter/pull/315 - as there is a lot of refactoring and then we can rebase this one and merge it.

Could take a while but a fan of your PR. Thank you!

Still I need a short note in the README.md that using the code as lib is allowed but not supported.
I won't resolve problems but will accept PRs in terms of yace as lib.
Changes can be breaking between releases and there won't be relase notes for people using the exporter as lib.
I hope you can understand that.

Greetings, Thomas

**Comment by jeschkies** [comment]
Great news. I think it's totally fine not to support YACE as a lib. I do like the pattern of `cmd` and `pkg`, though. I will try to merge in #315.

**Comment by thomaspeitz** [comment]
Could you please rebase? :)

**Comment by thomaspeitz** [comment]
Awesome. Nice. I will merge this one with the next release. I do not want two big changes in one release. That would make it harder to find errors.

So this will be merged but with next release (hope to release tomorrow) and then next week friday your release.

**Comment by jeschkies** [comment]
Nice. Thanks a lot!

**Comment by thomaspeitz** [comment]
Could you please rebase. Will try to merge it this week. Thanks for your time!

**Comment by jeschkies** [comment]
@thomaspeitz I've merged upstream.

**Comment by jeschkies** [comment]
Ping @thomaspeitz

**Comment by antoniomerlin** [comment]
In Dockerfile this need to be updated
`COPY ./*.go ./config_test.yml ./` 
to
`COPY ./*.go ./pkg/config_test.yml ./`

**Comment by jeschkies** [comment]
Thanks @antoniomerlin! I've created a follow up https://github.com/ivx/yet-another-cloudwatch-exporter/pull/359

**Pull request merged**